### ISSUE TITLE
Check if device is already mounted before mounting it again

### ISSUE THREAD

**Issue by probonopd** [vermaden/automount#37]
On the helloSystem Live ISO, some volumes (e.g., `MONKEYPATCH` in the picture) need to be mounted very early in the boot process, before `automount` is started. This results to these being mounted twice (see `MONKEYPATCH-2`).

It would be good to check if a geom is already mounted and if so, exit without trying to mount it again.

![image](https://user-images.githubusercontent.com/2480569/142064059-23063823-6646-41bb-a3cc-834e37120985.png)

**Comment by probonopd** [comment]
Note that `MONKEYPATCH` can be addressed as `/dev/da1` and `/dev/msdosfs/MONKEYPATCH`.

In this case, `/dev/msdosfs/MONKEYPATCH` was already mounted before `automount` was started, and then was mounted again from `/dev/da1`.

```
% mount
/dev/iso9660/LIVE on / (cd9660, local, read-only)
devfs on /dev (devfs)
tmpfs on /tmp (tmpfs, local)
tmpfs on /media (tmpfs, local)
/dev/md0.uzip on /media/uzip (ufs, local, read-only)
/media/uzip/System on /System (nullfs, local, read-only)
/media/uzip/Applications on /Applications (nullfs, local, read-only)
/media/uzip/bin on /bin (nullfs, local, read-only)
/media/uzip/lib on /lib (nullfs, local, read-only)
/media/uzip/libexec on /libexec (nullfs, local, read-only)
tmpfs on /nvidia (tmpfs, local)
/media/uzip/sbin on /sbin (nullfs, local, read-only)
/media/uzip/usr on /usr (nullfs, local, read-only)
/media/uzip/boot on /boot (nullfs, local, read-only)
tmpfs on /root (tmpfs, local)
tmpfs on /var (tmpfs, local)
tmpfs on /usr/local (tmpfs, local)
<below>:/media/uzip/usr/local on /usr/local (unionfs, local)
tmpfs on /etc (tmpfs, local)
/tmp/etc on /etc (nullfs, local)
tmpfs on /home (tmpfs, local)
<above>:/tmp/unionfs/Applications on /Applications (unionfs, local)
/dev/msdosfs/MONKEYPATCH on /media/MONKEYPATCH (msdosfs, local)
/dev/da0 on /media/LIVE (cd9660, local, read-only)
/dev/ada1p1 on /media/msdosfs (msdosfs, local, noatime)
/dev/fuse on /media/Wiederherstellung (fusefs, noatime, read-only)
/dev/fuse on /media/ntfs (fusefs, noatime, read-only)
/dev/ada0p2 on /media/msdosfs-2 (msdosfs, local, noatime)
map -hosts on /net (autofs)
/dev/da0p2 on /media/EFISYS (msdosfs, local, noatime, read-only)
/dev/da1 on /media/MONKEYPATCH-2 (msdosfs, local, noatime, read-only)
```

**Comment by vermaden** [comment]
Thats the 'price' for alternative approach with 'read only' mount.

How would you know if a device is mounted? Check output of **mount(8)** command.

After **/dev/msdosfs/MONKEYPATCH** was mounted ... does **/dev/da1** appear in the mounted list? No. So its not mounted to the automount. Then automount tries to mount it read write and it fails because its already mounted so it tries to mount it read only and that works.

Can you do not use **/dev/msdosfs** path for mounting and just use 'plain' devices as **da0/da1/...**?

**Comment by probonopd** [comment]
We can't trust that user would always do this, so we need to find a solution.

So I propose, when e.g., `/usr/local/sbin/automount da1 attach` is invoked, to add something like

```
dmesg | grep "Label for provider da1 is" | tail -n 1 | cut -d " " -f 7 | rev | cut -c 2- | rev
```

to `__check_already_mounted()`.

**Comment by vermaden** [comment]
That would also require additional checks for UFS labels, GPT labels, GLABEL labels, ext2/3/4 labels, and so on. It will make supporting automount(8) a nightmare ...

**Comment by probonopd** [comment]
Would you consider it if I come up with a grep one-liner?

**Comment by vermaden** [comment]
One grep(1) one liner would not solve that. You have to use multiple utilities to get the labels of many types of filesystems.

Just do not use **/dev/msdosfs** as the path for **automount(8)** daemon.

Use **da(4)** and **ada(4)** and **md(4)** and **mmcsd(4)** devices.

**Comment by probonopd** [comment]
That does not work for me. The volume with this label may be on different types of disks. I want to address my volumes by their name, not depending on the bus they happen to be connected to.

**Comment by vermaden** [comment]
I will try to add that 'check' for label at least for FAT fs.

That would be sufficient for your needs, yes?

**Comment by probonopd** [comment]
Unfortunately no. Do you know what else there is besides `msdosfs`, `iso9660`, `ufs`, `uzip` or where to look that up? I haven't given up on the idea of an one-liner entirely yet and would like to try.

**Comment by vermaden** [comment]
Also **/dev/gpart** and **/dev/label** for sure.

**Comment by probonopd** [comment]
Relevant docs:

https://docs.freebsd.org/en/books/handbook/geom/#geom-glabel

> By permanently labeling the partitions on the boot disk, the system should be able to continue to boot normally, even if the disk is moved to another controller or transferred to a different system.

This is the reason why we want to use the disk labels for everything in helloSystem.

**Comment by probonopd** [comment]
Looking at `[man glabel(8)](https://www.freebsd.org/cgi/man.cgi?query=glabel&sektion=8&format=html)` it seems like we need to check:

- /dev/label/
- /dev/ufs/
- /dev/ufsid/
- /dev/msdosfs/
- /dev/iso9660/
- /dev/ext2fs/
- /dev/reiserfs/
- /dev/ntfs/
- /dev/gpt/
- /dev/gptid/
- /dev/diskid/

**Comment by vermaden** [comment]
Yeah, that is why I stick to the 'raw' devices and not their logical clones.

I have other idea. I will add an option to the **automount.conf** config file that will disallow read only mounts - do each volume will be again mounted once.

**Comment by probonopd** [comment]
`mount | grep -e "^/dev/(label|ufs|ufsid|msdosfs|iso9660|ext2fs|reiserfs|ntfs|gpt|gptid|diskid)/MONKEYPATCH on"`

**Comment by probonopd** [comment]
>  I will add an option to the automount.conf config file that will disallow read only mounts 

Who  would want that? I will need read-only mounts _a lot_ for sure, especially that now double-clicking an ISO file mounts it!

**Comment by vermaden** [comment]
But WHERE that **grep(1)** expression would help us?

**Comment by probonopd** [comment]
Before doing any mounting, run
`<that grep expression> && exit 0`

**Comment by vermaden** [comment]
Check with latest version.

- https://twitter.com/vermaden/status/1694440196236849227
- https://bsd.network/@vermaden/110940718791974585
- https://mastodon.bsd.cafe/@vermaden/110940719515574245

Thanks.
